Library/Homebrew/dev-cmd/bump-formula-pr.rb

```diff
@@ -35,8 +35,7 @@ def bump_formula_pr_args
       switch "-n", "--dry-run",
              description: "Print what would be done rather than doing it."
       switch "--all",
```

Comment:
```suggestion
      switch "--guess-formula",
```

Though feels like a weird feature in general to be honest to know the URL you want to update but not the formula.

Reply:
@Bo98 Specifically want anywhere that does `Formula.all` to require `--all`. Consistency seems important for a security feature like this. The alternative is we rip out this code entirely.

Reply:
In terms of ripping out stuff: I'm fine with this happening for `bump-formula-pr` unless you can find someone who uses this.

> Specifically want anywhere that does Formula.all to require `--all`.

Why? It's confusing when you have something like `brew bump-formula-pr --all` that doesn't make any sense name wise. I'd argue confusion is detrimental to security rather than benefitting it.

Reply:
> Why? It's confusing when you have something like brew bump-formula-pr --all that doesn't make any sense name wise. I'd argue confusion is detrimental to security rather than benefitting it.

Because it allows us to have a security check like the one that's been commented out in `Formula.all`. It also teaches users "if you want to run a command that will read all formulae/casks you have installed: you need to pass `--all`". 

I agree this descriptions can be improved and aliases added but I don't agree that having a collection of arguments which will read formulae will provide the same benefit.

Reply:
> Because it allows us to have a security check like the one that's been commented out in `Formula.all`.

Feels like arg bleed, particularly since that can be used out of the context of commands. If you want to block it outside of commands, it sounds you like you want to deprecate it completely and introduce a private API instead. Or potentially even something more in CLI parser land.

> It also teaches users "if you want to run a command that will read all formulae/casks you have installed: you need to pass `--all`".

An example where this already doesn't make sense is `brew readall`. The `--all` flag makes sense there but does have a different meaning: read all taps rather than a single tap. It still reads all formulae in the specified taps regardless.

Reply:
> Or potentially even something more in CLI parser land.

We've avoided a global CLI parsing instance in the past. We could potentially have something like `Homebrew.all_passed?` but not sure if that would be better.

> introduce a private API instead. 

The issue is not whether the API is public or private but how it is used.

> It's all very well that bump-formula-pr can read all formulae optionally but that doesn't say anything about what it's actually doing that for.

I don't think that really matters from a security perspective. You're either executing arbitrary code from all formulae and all taps or you're only acting upon specific formulae requested by name (or by that formula).

> Discouraging the use of the flag while encouraging the people to copy paste the --all flag everywhere in their scripts without consideration seems conflicting.

I don't think this is being encouraged. Yeh, we have this flag for this reason. Setting the flag is a nuclear option, though. I don't think it should be required for literally every use; that feels obnoxious.

I also don't think a "do this at all times for all commands without prompting" is a better thing to be nudging people towards than "use `--all` for this specific command".

Reply:
> We've avoided a global CLI parsing instance in the past.

Exactly, which is why all the arg checking is mostly gone outside of `(dev-)cmd/*`. `build.rb` I think is the only straggler.

> The issue is not whether the API is public or private but how it is used.

This is a relevant question for third-party scripts. If you don't want to allow it to ever be used outside of something tied to a CLI parser, maybe `Formula` is now the wrong place to put it.

(Note that Homebrew/command-not-found for one needs updating)

> I don't think that really matters from a security perspective. You're either executing arbitrary code from all formulae and all taps or you're only acting upon specific formulae requested by name (or by that formula).

So there's no reason to have per-command flags?

The point of the per-command flags I thought is so you can have scoped decisions on when you need to do this, which requires knowing the purpose of executing the said arbitrary code. So the "why" seems relevant to me.

> I also don't think a "do this at all times for all commands without prompting" is a better thing to be nudging people towards than "use --all for this specific command".

Absolutely not - I think deciding on a per-command basis is a good thing. But I don't see how that's a reason for `--all` to be everywhere if you want people to check each individual command?

Reply:
> Exactly, which is why all the arg checking is mostly gone outside of (dev-)cmd/*. build.rb I think is the only straggler.

Yeh, note the `ARGV` check here isn't even live code. It's something we can figure out for 3.7.0.

> If you don't want to allow it to ever be used outside of something tied to a CLI parser, maybe Formula is now the wrong place to put it.

Yes, that's fair. I'm not sure where best to put it, though? On the parser itself?

> The point of the per-command flags I thought is so you can have scoped decisions on when you need to do this, which requires knowing the purpose of executing the said arbitrary code. So the "why" seems relevant to me.

Yeh, the why can be reflected on the descriptions. I'm happy to take suggestions on how to improve them.

> Absolutely not - I think deciding on a per-command basis is a good thing. But I don't see how that's a reason for --all to be everywhere if you want people to carefully consider and check each individual command?

Consistency seems valuable here. I'm open to it being not-just `--all`, though, it could be `--all-dangerous` or something.

> The only reason I can see for it being uniform is the deprecation code - which makes sense but I think it can be done in a better way API wise (see above).

I still think done in a better way will be easier to follow both internally and externally if using a consistent flag, even if we don't use `ARGV` to enforce this.

Reply:
> Yes, that's fair. I'm not sure where best to put it, though? On the parser itself?

It makes sense to me on the surface that something that requires a CLI parser is attached to something in the CLI code (so one of the three CLI components there).

> Consistency seems valuable here. I'm open to it being not-just --all, though, it could be --all-dangerous or something.

While I might still not follow the consistency value here given we want people to read manpages, I will say if it's something that has to appear everywhere then something less of a common-sounding regular flag (if that makes sense?) than `--all` does make sense. So yeah, something like `--all-dangerous` or whatever.

Reply:
`--read-all`, `--readall`, `--execall`? `--force-all`? Want to balance being short with being explanatory.

Reply:
Not sure what's best but out of those `--exec-all` is one of the better ones. I try to think of them in context of the commands in question. For example `brew readall --readall` sounds off.

Reply:
I like `--exec-all` a lot more than `--all` (but would probably be okay with `--all` or `-a` being aliases)

To me, `brew readall --all`, `brew tap --all`, `brew audit --all` and `brew bump-formula-pr --all` are very confusing

Another option might be to add this flag to the [global options](https://github.com/Homebrew/brew/blob/39d9f14dc8efb353ac9b3abe5efadd708bd7fb5b/Library/Homebrew/cli/parser.rb#L101-L108), although that probably isn't necessary